src/DotnetLearn.QueryFilterDemo/QueryFilterContext.cs

```diff
@@ -1,31 +1,87 @@
+        ) : base(options)
+    {
+        // this.tenantId = tenantId;
+    }
+
     public DbSet<Order> Orders { get; set; } = null!;
     public DbSet<Customer> Customers { get; set; } = null!;
     public DbSet<Tenant> Tenants { get; set; } = null!;
 
-    public QueryFilterContext(DbContextOptions options) : base(options) { }
-
     protected override void OnModelCreating(ModelBuilder modelBuilder)
     {
         var order = modelBuilder.Entity<Order>();
-        order.HasAlternateKey(order => (new { order.Id, order.TenantId }));
-        order.HasOne(order => order.Tenant)
-            .WithMany()
-            .HasForeignKey(order => order.TenantId);
-        order.HasOne(order => order.Customer)
-            .WithMany()
-            .HasForeignKey(order => new { order.CustomerId, order.TenantId })
-            .HasPrincipalKey(customer => new { customer.Id, customer.TenantId });
-
         var customer = modelBuilder.Entity<Customer>();
-        customer.HasAlternateKey(customer => (new { customer.Id, customer.TenantId }));
-        customer.HasOne(customer => customer.Tenant)
-            .WithMany()
-            .HasForeignKey(customer => customer.TenantId);
+        var tenant = modelBuilder.Entity<Tenant>();
+
+        ConfigureOrder(order);
+        ConfigureCustomer(customer);
+
+        static void ConfigureCustomer(EntityTypeBuilder<Customer> customer)
+        {
+            customer.HasAlternateKey(customer => (new { customer.Id, customer.TenantId }));
+            customer.HasOne(customer => customer.Tenant)
+                .WithMany()
+                .HasForeignKey(customer => customer.TenantId);
+        }
+
+        static void ConfigureOrder(EntityTypeBuilder<Order> order)
+        {
+            order.HasAlternateKey(order => (new { order.Id, order.TenantId }));
+            order.HasOne(order => order.Tenant)
+                .WithMany()
+                .HasForeignKey(order => order.TenantId);
+            order.HasOne(order => order.Customer)
+                .WithMany()
+                .HasForeignKey(order => new { order.CustomerId, order.TenantId })
+                .HasPrincipalKey(customer => new { customer.Id, customer.TenantId });
+        }
+
+        // tenant.AddSoftDeleteQueryFilter();
+        // customer.AddSoftDeleteQueryFilter();
+        // order.AddSoftDeleteQueryFilter();
+
+        // modelBuilder.AddSoftDeleteQueryFilters();
+
+        // tenant.HasQueryFilter(tenant => tenant.Id == tenantId);
+        // customer.HasQueryFilter(customer => customer.TenantId == tenantId);
+        // order.HasQueryFilter(order => order.TenantId == tenantId);
+
+        // tenant.AddQueryFilter(tenant => tenant.Id == tenantId);
+        // customer.AddQueryFilter(customer => customer.TenantId == tenantId);
+        // order.AddQueryFilter(order => order.TenantId == tenantId);
+
+        // modelBuilder.AddTenantQueryFiltersBad(tenantId);
+
+        // modelBuilder.AddTenantQueryFilters(this);
```

Comment:
This are the incremental changes that we'll mainly be demoing

First adding soft delete query filters by hand
Then reflecting and detecting soft deleteables by the marker interface
Then adding tenant id query filters, works but overrides the soft delete filter
Then combining query filters
Then reflecting to config on all tenant specific entities by, doesn't work and (caches tenantId between contexts)
Fix caching